remote-trait-object/src/service/serde_support.rs

```diff
@@ -20,65 +20,32 @@ use parking_lot::RwLock;
 use serde::de::{Deserialize, Deserializer};
 use serde::ser::{Serialize, Serializer};
 
-pub struct SBox<T: ?Sized + Service> {
-    value: std::cell::Cell<Option<Box<T>>>,
+pub struct ServicePointer<T, P: ?Sized + Service> {
+    value: std::cell::Cell<Option<T>>,
+    /// This is for the generic Serialize/Deserialize implementation.
+    _p: std::marker::PhantomData<P>,
 }
 
-impl<T: ?Sized + Service> SBox<T> {
-    pub fn new(value: Box<T>) -> Self {
-        SBox {
+impl<T, P: ?Sized + Service> ServicePointer<T, P> {
+    pub fn new(value: T) -> Self {
+        ServicePointer {
             value: std::cell::Cell::new(Some(value)),
+            _p: std::marker::PhantomData,
         }
     }
 
-    pub(crate) fn take(&self) -> Box<T> {
+    pub(crate) fn take(&self) -> T {
         self.value.take().unwrap()
     }
 
-    pub fn unwrap(self) -> Box<T> {
+    pub fn unwrap(self) -> T {
         self.value.take().unwrap()
     }
 }
 
-pub struct SArc<T: ?Sized + Service> {
-    value: std::cell::Cell<Option<Arc<T>>>,
-}
-
-impl<T: ?Sized + Service> SArc<T> {
-    pub fn new(value: Arc<T>) -> Self {
-        SArc {
-            value: std::cell::Cell::new(Some(value)),
-        }
-    }
-
-    pub(crate) fn take(&self) -> Arc<T> {
-        self.value.take().unwrap()
-    }
-
-    pub fn unwrap(self) -> Arc<T> {
-        self.value.take().unwrap()
-    }
-}
-
-pub struct SRwLock<T: ?Sized + Service> {
-    value: std::cell::Cell<Option<Arc<RwLock<T>>>>,
-}
-
-impl<T: ?Sized + Service> SRwLock<T> {
-    pub fn new(value: Arc<RwLock<T>>) -> Self {
-        SRwLock {
-            value: std::cell::Cell::new(Some(value)),
-        }
-    }
-
-    pub(crate) fn take(&self) -> Arc<RwLock<T>> {
-        self.value.take().unwrap()
-    }
-
-    pub fn unwrap(self) -> Arc<RwLock<T>> {
-        self.value.take().unwrap()
-    }
-}
+pub type SBox<T> = ServicePointer<Box<T>, T>;
+pub type SArc<T> = ServicePointer<Arc<T>, T>;
+pub type SRwLock<T> = ServicePointer<Arc<RwLock<T>>, T>;
```

Comment:
I think the previous version is easier to understand.
Why did you change this code?